File: src/prismeai/resources/a2a.py
from __future__ import annotations
from typing import Any, TYPE_CHECKING
from .._streaming import Stream, AsyncStream
if TYPE_CHECKING:
    from .._client import SyncAPIClient, AsyncAPIClient

_req_id = 0

def _next_id() -> str:
    global _req_id
    _req_id += 1
    return f"req-{_req_id}"

class A2A:
    def __init__(self, client: SyncAPIClient, workspace_id: str) -> None:
        self._client = client
        self._workspace_id = workspace_id
    def _path(self, *segments: str) -> str:
        parts = "/".join(segments)
        return f"/workspaces/slug:{self._workspace_id}/webhooks/v1/{parts}"
    def send(self, agent_id: str, **kwargs: Any) -> Any:
        return self._client.post(self._path("agents", agent_id, "a2a"), json={
            "jsonrpc": "2.0",
            "method": "tasks/send",
            "params": kwargs,
            "id": _next_id(),
        })
    def send_subscribe(self, agent_id: str, **kwargs: Any) -> Stream[Any]:
        response = self._client.request("POST", self._path("agents", agent_id, "a2a"), json={
            "jsonrpc": "2.0",
            "method": "tasks/sendSubscribe",
            "params": kwargs,
            "id": _next_id(),
        }, headers={"accept": "text/event-stream"}, raw=True)
        return Stream(response)
    def get_card(self, agent_id: str) -> Any:
        return self._client.get(self._path("agents", agent_id, ".well-known", "agent.json"))
    def get_extended_card(self, agent_id: str) -> Any:
        return self._client.get(self._path("agents", agent_id, "extendedAgentCard"))

class AsyncA2A:
    def __init__(self, client: AsyncAPIClient, workspace_id: str) -> None:
        self._client = client
        self._workspace_id = workspace_id
    def _path(self, *segments: str) -> str:
        parts = "/".join(segments)
        return f"/workspaces/slug:{self._workspace_id}/webhooks/v1/{parts}"
    async def send(self, agent_id: str, **kwargs: Any) -> Any:
        return await self._client.post(self._path("agents", agent_id, "a2a"), json={
            "jsonrpc": "2.0",
            "method": "tasks/send",
            "params": kwargs,
            "id": _next_id(),
        })
    async def send_subscribe(self, agent_id: str, **kwargs: Any) -> AsyncStream[Any]:
        response = await self._client.request("POST", self._path("agents", agent_id, "a2a"), json={
            "jsonrpc": "2.0",
            "method": "tasks/sendSubscribe",
            "params": kwargs,
            "id": _next_id(),
        }, headers={"accept": "text/event-stream"}, raw=True)
        return AsyncStream(response)
    async def get_card(self, agent_id: str) -> Any:
        return await self._client.get(self._path("agents", agent_id, ".well-known", "agent.json"))
    async def get_extended_card(self, agent_id: str) -> Any:
        return await self._client.get(self._path("agents", agent_id, "extendedAgentCard"))

File: src/resources/agents/a2a.ts
import { BaseResource } from '../../core/base-resource.js';
import type { HttpClient } from '../../core/http-client.js';
import { SSEStream } from '../../core/streaming.js';
import type {
  A2ASendParams,
  A2AResponse,
  A2ACard,
  A2AExtendedCard,
  A2AStreamEvent,
} from '../../types/a2a.js';

let _reqId = 0;
function nextId(): string {
  return `req-${++_reqId}`;
}

export class A2A extends BaseResource {
  private readonly workspaceId: string;

  constructor(httpClient: HttpClient, workspaceId: string) {
    super(httpClient);
    this.workspaceId = workspaceId;
  }

  private path(...segments: string[]): string {
    return this.buildPath(this.workspaceId, ...segments);
  }

  /** Send a message via A2A protocol (JSON-RPC 2.0). */
  send(agentId: string, params: A2ASendParams): Promise<A2AResponse> {
    return this._post<A2AResponse>(this.path('agents', agentId, 'a2a'), {
      jsonrpc: '2.0',
      method: 'tasks/send',
      params,
      id: nextId(),
    });
  }

  /** Send a message and subscribe to SSE events (A2A protocol, JSON-RPC 2.0). */
  async sendSubscribe(agentId: string, params: A2ASendParams): Promise<SSEStream<A2AStreamEvent>> {
    const controller = new AbortController();
    const response = await this.httpClient.requestRaw({
      method: 'POST',
      path: this.path('agents', agentId, 'a2a'),
      body: {
        jsonrpc: '2.0',
        method: 'tasks/sendSubscribe',
        params,
        id: nextId(),
      },
      headers: {
        accept: 'text/event-stream',
      },
      signal: controller.signal,
    });

    return new SSEStream<A2AStreamEvent>(response, controller);
  }

  /** Get the A2A agent card (.well-known/agent.json). */
  getCard(agentId: string): Promise<A2ACard> {
    return this.httpClient.get<A2ACard>(this.path('agents', agentId, '.well-known', 'agent.json'));
  }

  /** Get the extended A2A agent card with tools and sub-agents. */
  getExtendedCard(agentId: string): Promise<A2AExtendedCard> {
    return this.httpClient.get<A2AExtendedCard>(this.path('agents', agentId, 'extendedAgentCard'));
  }
}
